fedbiomed/node/secagg_dummy_setup.py

```diff
@@ -0,0 +1,119 @@
+        self._experiment_id = experiment_id
+        self._parties = parties
+        self._grpc_client = grpc_client
+        self._pending_requests = pending_requests
+
+    def _create_secagg_reply(self, message: str = '', success: bool = False) -> dict:
+        """Same as `BaseSecaggSetup._create_secagg_reply()`
+
+        TODO: use original function when refactoring `SecaggSetup`
+        """
+
+        # If round is not successful log error message
+        if not success:
+            logger.error(message)
+
+        return {
+            'researcher_id': self._researcher_id,
+            'secagg_id': self._secagg_id,
+            'success': success,
+            'msg': message,
+            'command': 'secagg'
+        }
+
+    def setup(self) -> dict:
+        """Set up a secagg context element.
+
+        Returns:
+            message to return to the researcher after the setup
+        """
+
+        other_nodes = [ e for e in self._parties[1:] if e != environ['NODE_ID'] ]
+
+        # Key exchange with other nodes
+
+        other_nodes_messages = []
+        for node in other_nodes:
+            other_nodes_messages += [
+                NodeToNodeMessages.format_outgoing_message({
+                    'request_id': REQUEST_PREFIX + str(uuid.uuid4()),
+                    'node_id': environ['NODE_ID'],
+                    'dest_node_id': node,
+                    'dummy': f"KEY REQUEST INNER from {environ['NODE_ID']}",
+                    'secagg_id': self._secagg_id,
+                    'command': 'key-request'
+                })
+            ]
+        logger.debug(f'Sending Diffie-Hellman setup for {self._secagg_id} to nodes: {other_nodes}')
+        try:
+            listener_id = send_overlay_message(
+                self._grpc_client,
+                self._pending_requests,
+                self._researcher_id,
+                other_nodes,
+                other_nodes_messages,
+            )
+
+            all_received, messages = self._pending_requests.wait(listener_id, TIMEOUT_NODE_TO_NODE_REQUEST)
+
+        # TODO: for real use, this code will be in `BaseSecaggSetup._setup_specific()`
+        # which catches FedbiomedError exceptions.
+        except FedbiomedError as e:
+            logger.debug(f"{e}")
+            return self._create_secagg_reply(f'Can not setup secure aggregation it might be due to unregistered '
+                                             f'certificate for the federated setup. Please see error: {e}', False)
+        except Exception as e:
+            logger.debug(f"{e}")
+            return self._create_secagg_reply('Unexpected error occurred please '
+                                             'report this to the node owner', False)
+
+        logger.debug(f'Completed Diffie-Hellmann setup for {self._secagg_id}. Status: {all_received}')
+
+        return self._create_secagg_reply(
+            "Secagg element context setup for Diffie-Hellman completed", all_received
+        )
```

Comment:
If you agree @srcansiz I would propose to address this point in #1135 where we have the final code for `SecaggDhSetup`. In this PR we just sketch it.

I might be missing something but I don't see what would be the difficulty to fuse `send_nodes()` and `wait()`. I implemented like this because it was our initial discussion. We need an argument to specifiy `wait=True`, as you mentioned because I also implement the case of of non-request-reply messages (single messages) where we don't want to wait for replies at all.

Reply:
Sorry maybe I was not explicit in the example usage that shared. (1) My point was `send_overlay_message` could be implemented in a way that we don't need to pass `grpc_client` and `pending_request` each time we need to send a request. (2) As developer the naming of the method is not explicit (`send_overlay_message`), I prefer a naming that explicitly indicates that this methods sends messages to nodes. (3) calling `wait` is not difficult but it can be simpler if it is called automatically from  the `send_overlay_message`. (4) as a developer, I do not prefer catch unexpected exceptions each time I call `send_overlay_message`, instead I prefer `send_overlay_message` handles unexpected exceptions and I can catch only `FedbiomedError` to return a status.

Reply:
Thank you for clarification. 

Would you agree @srcansiz  to move this comment to #1135 where the API for this part is supposed to get finalized ? (to avoid making effort now while unsure whether it conflct with further modification)

Reply:
Yes yes sure :)